### PR TITLE
prevent take() from requesting more than needed

### PR DESCRIPTION
**Pull request by davidmoten**
The `request` method used logic that could overrequest upstream if multiple small requests were made before the next item arrived.

Added a unit test that failed on previous code.

### CODE REVIEW

File: src/main/java/rx/internal/operators/OperatorTake.java [hunk 1]
```diff
@@ -78,15 +80,22 @@ public void onNext(T i) {
...
-                            if (n < c) {
-                                producer.request(n);
-                            } else {
-                                producer.request(c);
+                        if (n >0 && !completed) {
+                            // because requests may happen concurrently use a CAS loop to 
+                            // ensure we only request as much as needed, no more no less
+                            while (true) {
+                                long r = requested.get();
+                                long c = Math.min(n, limit - requested.get());
```
Comment (akarnokd) [inline src/main/java/rx/internal/operators/OperatorTake.java]: Re-read of a volatile value, you need ```limit - r```.

File: src/main/java/rx/internal/operators/OperatorTake.java [hunk 2]
```diff
@@ -78,15 +80,22 @@ public void onNext(T i) {
...
+                                if (requested.compareAndSet(r, r + c)) {
```
Comment (akarnokd) [inline src/main/java/rx/internal/operators/OperatorTake.java]: if ```c == 0``` there is nothing to request and the loop can quit.

**Comment by davidmoten** [comment]
Thanks @akarnokd. Changes made.

**Comment by akarnokd** [comment]
Great job!

**Pull request merged**